javascript/i18n/phonenumbers/asyoutypeformatter.js

```diff
@@ -354,16 +358,30 @@ i18n.phonenumbers.AsYouTypeFormatter.prototype.getAvailableFormats_ =
   for (var i = 0; i < formatListLength; ++i) {
     /** @type {i18n.phonenumbers.NumberFormat} */
     var format = formatList[i];
-    /** @type {boolean} */
-    var nationalPrefixIsUsedByCountry =
-        this.currentMetadata_.hasNationalPrefix();
-    if (!nationalPrefixIsUsedByCountry || this.isCompleteNumber_ ||
-        format.getNationalPrefixOptionalWhenFormatting() ||
+    // Discard a few formats that we know are not relevant based on the
+    // presence of the national prefix.
+    if (this.extractedNationalPrefix_.length > 0 &&
         this.phoneUtil_.formattingRuleHasFirstGroupOnly(
-            format.getNationalPrefixFormattingRuleOrDefault())) {
-      if (this.isFormatEligible_(format.getFormatOrDefault())) {
-        this.possibleFormats_.push(format);
-      }
+            format.getNationalPrefixFormattingRuleOrDefault()) &&
+        !format.getNationalPrefixOptionalWhenFormatting() &&
+        !format.hasDomesticCarrierCodeFormattingRule()) {
+      // If it is a national number that had a national prefix, any rules that
+      // aren't valid with a national prefix should be excluded. A rule that
+      // has a carrier-code formatting rule is kept since the national prefix
+      // might actually be an extracted carrier code - we don't distinguish
+      // between these when extracting it in the AYTF.
+      continue;
+    } else if (this.extractedNationalPrefix_.length == 0 &&
+        !this.isCompleteNumber_ &&
+        !this.phoneUtil_.formattingRuleHasFirstGroupOnly(
+            format.getNationalPrefixFormattingRuleOrDefault()) &&
+        !format.getNationalPrefixOptionalWhenFormatting()) {
+      // This number was entered without a national prefix, and this formatting
+      // rule requires one, so we discard it.
+      continue;
+    }
+    if (this.isFormatEligible_(format.getFormatOrDefault())) {
```

Comment:
maybe this could be inlined ... not sure

Reply:
yep fixed.